### ISSUE TITLE
Attempt::get_workout_score is broken

### ISSUE THREAD

**Issue by s-edwards** [web-cat/code-workout#14]
The code in this method assumes that all of a user's attempts in a workout are stored sequentially by id, and that they are contiguous with no other user's attempts in between.  This is plainly flawed, and the method needs to be rewritten.  Overall, it looks like the whole attempts infrastructure needs to be rewritten :-(.

**Comment by kbuffardi** [comment]
@pmk1433 I believe when I created the attempts model I included fields for both the **time** an attempt is submitted and also a field that (auto?) incrementally **counts** the number of attempts made (so the row could tell you "User X's fifth attempt of Exercise Y was submitted at 12:34.00..."). If you  query by username and sort by **submit_num**, you should have the order of each attempt by that user.

**Comment by s-edwards** [comment]
@kbuffardi, yes, the submit_num column can be used to retrieve the most recent attempt.  However, the problem is a bit deeper.  The given method (which has a poor name) appears to be trying to calculate the total score for an entire workout.  Yes, its approach to retrieving all the attempts for all the exercises is broken.

Currently, however, there's no way to determine if an attempt was completed as part of a workout or not in the model, and there is no effective way to find all the "final" attempts for all the exercises in a workout so far.  Remember that an exercise might have been practiced before a workout was assigned, or might have been re-practiced after a workout was completed and given a score.  As a result, just looking at the most recent attempt won't necessarily give the correct results anyway.

**Comment by kbuffardi** [comment]
Ah, good point. I hadn't previously considered how attempts would calculate scores separately between a practiced exercise and a workout exercise.

My initial thought is that we could add a relationship (belongs_to) between an attempt and a workout, which would indicate if it's an attempt from an exercise within a workout. If we do that we'll also have to reset the submit_num to 1 when doing it as part of a workout so there might be submit_num=1 for pre-workout practice and also a submit_num=1 as part of the workout. The final (or at least last attempt) of each exercise in a workout could then be found by querying user's attempts for an exercise WITH that workout id SORTED by submit_num descending.

With this approach, we'd have to make sure to...
(1) verify (via session data?) whether an exercise being attempted is part of an active workout (before the workout's due date and/or came to the exercise directly from a workout link), 
(2) assign the relationship (or lack thereof) to the workout
(3) (re)set/increment submit_num accordingly -- each affiliated workout would have its own count

This would require rewriting the method, but doesn't require many changes to the Attempt model itself. 

However, things can get a little ugly if an exercise belongs to multiple workouts that are active at the same time. Say a student has two different workouts assigned at the same time and they share a common exercise. Does that mean the student should have to perform the exercise twice? Or does she just have to do it once and it'll count toward both workouts? The latter approach wouldn't work with a simple belongs_to relationship in which case we'd have to consider other solutions.

**Comment by pmk1433** [comment]
@s-edwards @kboof I have finally figured out what I did for this broken method and why it was that I did what I did. 

tl; dr I had recognized the 'brokeness' of this approach and had created the WorkoutScore model that has the proper mechanism as described above (but only if the workout was part of an offering, so that assumption is still a mistake). This made the method in question obsolete but I didn't want to delete this method and so I left it hanging.

Around the middle of January, when I was working recklessly to develop some features, I noticed that I could not consistently get information that I wanted about a user's score in a workout and I developed this method in a hurry. After repeated testing of that method with test attempts, I had realized the different problems inherent with the above method as @s-edwards had described above. 

So, I created the WorkoutScore model and added some mechanisms to it so that for each workout/user pair, the record_workout_score will correctly record the score for its component exercises (but only if the workout was part of an offering otherwise the logic is incorrect). Note that I let the record_workout_score method hang around in CodeWorker because at that I was unsure about whether it would threaded properly. The record_workout_score should probably refactored so that it is a class method of WorkoutScore, per OOP paradigm. Probably the same for the record_attempt, I guess.

Anyways, the method that is the focus of this issue was made unnecessary but I could not bring myself to delete it. I didn't recognize what had happened in this regard in February when this issue was raised because the title had a different method name. Apologies for my mistake.

I think we can build a better robust mechanism by augmenting the Attempt model with a workout_id column and slightly modify the record_workout_score so that it will different between attempts that were a part of a workout and those that were not.

I will take care of this once we have ironed out the issues.

**Comment by pmk1433** [comment]
Removed the faulty method.
